### ISSUE TITLE
C++14 / 11 compatibility

### ISSUE THREAD

**Issue by timcardenuto** [RedhawkSDR/redhawk#3]
**TL;DR**  Made a few minor patches to build Redhawk 2.0.7 for Debian 9 and Raspbian Stretch posted [here](https://github.com/timcardenuto/core-framework/releases/tag/2.0.7-1). There were some issues related to the use of GCC 6.3 which uses C++14 by default, though some of these may be seen with C++11. Full process for Raspbian image is documented [here](https://gist.github.com/timcardenuto/8b497030795b07cdc681eb8074fd2003). Not real familiar with Issues/Pull Requests so let me know if there's anything standard I should be doing. 

List of modifications needed:
1. Error in `control/sdr/dommgr/applicationSupport.cpp`:

        applicationSupport.cpp: In member function 'bool ossie::ComponentInfo::isAssignedToDevice() const':
        applicationSupport.cpp:882:12: error: cannot convert 'const boost::shared_ptr<ossie::DeviceNode>' to 'bool' in return
             return assignedDevice;
             ^~~~~~~~~~~~~~~~~~~~~
             
    Fixed by following advice [here](https://stackoverflow.com/questions/7580009/gcc-error-cannot-convert-const-shared-ptr-to-bool-in-return/7580097#7580097), to replace the line `return assignedDevice;` with the syntax `return static_cast<bool> (assignedDevice);`.

2. Error in `testing/sdr/dev/devices/CppTestDevice/cpp/CppTestDevice.h`:

        CppTestDevice.h:31:33: error: 'constexpr' needed for in-class initialization of static data member 'const float CppTestDevice_i::MAX_LOAD' of non-integral type [-fpermissive]
            static const float MAX_LOAD = 4.0;
                                          ^~~
     Fixed by following advice [here](https://stackoverflow.com/questions/9141950/initializing-const-member-within-class-declaration-in-c), and replacing the line `static const float MAX_LOAD = 4.0;` with the line `static constexpr float MAX_LOAD = 4.0;`

3. Error in `control/framework/nodebooter.cpp` logging message at line 144:

        nodebooter.cpp:144:59: error: no match for ‘operator<<’ (operand types are ‘std::basic_ostream<char>’ and ‘std::ifstream {aka std::basic_ifstream<char>}’)
         LOG_ERROR(nodebooter, "Failed to parse PRF file " << prfStream<< ". " << parser_error_line << "The XML parser returned the following error: " << ex.what());
         
     Uses `prfStream` where `prfFile` string is needed. Also probably a mistake (doesn't throw any warnings/errors) in line 539 that references `dcdFile` but meant to use `dmdFile`.

4. Error in `testing/sdr/dom/components/msg_through_cpp/cpp/Makefile.am`, missing the `$(OMNICOS_LIBS)` linker flag to following line:
 
        msg_through_cpp_LDADD = $(CFDIR)/framework/libossiecf.la $(CFDIR)/framework/idl/libossieidl.la $(PROJECTDEPS_LIBS) $(BOOST_LDFLAGS) $(BOOST_THREAD_LIB) $(BOOST_REGEX_LIB) $(BOOST_SYSTEM_LIB) $(INTERFACEDEPS_LIBS) $(OMNICOS_LIBS) $(OMNIDYNAMIC_LIBS) $(redhawk_LDADD_auto)

5. Error in `GPP/cpp/GPP.cpp`:

        GPP.cpp:583:24: error: cannot bind 'std::basic_ostream<char>' lvalue to 'std::basic_ostream<char>&&'
        errstr << "Unable to read "<<stat_filename<<". The process is no longer there";
        ~~~~~~~^~~~~~~~~~~~~~~~~~~~
        
    Change `stat_filename` to `stat_filename.str()`.
 
6. Error in `bulkioInterfaces/libsrc/cpp/bulkio_out_port.cpp`:

        cpp/bulkio_out_port.cpp:500:94: error: no matching function for call to 'make_pair(BULKIO::StreamSRI&, bool)'
            ret[cSri->first] = std::make_pair< BULKIO::StreamSRI, bool >( cSri->second.sri, false );
 
   Change line 500 to `ret[cSri->first] = std::make_pair( cSri->second.sri, false );` which matches other use of make_pair and follows advice to remove explicit template [here](https://stackoverflow.com/questions/9641960/c11-make-pair-with-specified-template-parameters-doesnt-compile). 
   
7. Error in `bulkioInterfaces/libsrc/cpp/bulkio_attachable_port`:

        cpp/bulkio_attachable_port.cpp:1737:25:   required from ‘void bulkio::OutAttachablePort<StreamDefinition, PT, StreamSequence>::pushSRI(const BULKIO::StreamSRI&, const BULKIO::PrecisionUTCTime&) [with StreamDefinition = BULKIO::VITA49StreamDefinition; PT = BULKIO::dataVITA49; StreamSequence = BULKIO::VITA49StreamSequence]’
        cpp/bulkio_attachable_port.cpp:2142:18:   required from here
        /usr/include/c++/6/bits/stl_tree.h:1889:28: error: no matching function for call to ‘std::_Rb_tree<std::__cxx11::basic_string<char>, std::pair<const std::__cxx11::basic_string<char>, bulkio::AttachableSriMapStruct>, std::_Select1st<std::pair<const std::__cxx11::basic_string<char>, bulkio::AttachableSriMapStruct> >, std::less<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, bulkio::AttachableSriMapStruct> > >::_M_get_insert_unique_pos(std::pair<_CORBA_String_member, bulkio::AttachableSriMapStruct>::first_type&)’
          = _M_get_insert_unique_pos(_KeyOfValue()(__v));

    Change line 1737 to `currentSRIs.insert(std::make_pair(ossie::corba::returnString(H.streamID), sri_ctx));`

**Comment by awsaunier** [comment]
Can you be more specific with problem number 4.  I don't know where to add the flag.  Thank you for writing "Redhawk on Raspberry Pi".  It has really helped me moving in a direction and I look forward to getting this to run.

chears,
- Adam

**Comment by timcardenuto** [comment]
Added #4 clarification. Remember all these notes were for Redhawk version 2.0.7 so I can't vouch for how different it is now. I'm hoping to do an update for the 2.2.x release sometime in the next few months, I'm getting back into things with the PlutoSDR which has an ARM Cortex A9 somewhat similar to the old Pi's.

**Comment by shmahon** [comment]
Tracked by EXT-202.  Instead of making incremental modifications to the REDHAWK source code a section has been added to the REDHAWK [Manual](https://redhawksdr.github.io/2.2.5/manual/appendices/source-installation/#installing-the-framework-from-source) to explain how to use GCC flags to build REDHAWK on platforms with newer, default versions of the GCC libraries.

Closing this issue.
